### PR TITLE
added:

### PR DESCRIPTION
**Pull request by boxxxie**
Map.new_nested_map
Map.rename_keys (pruned, and unpruned versions)

### CODE REVIEW

**Comment by boxxxie** [comment]
@kenips this is useful for the import stuff that Robert is doing. will let us remove a lot of code

File: lib/kitchen-sink/map.ex [hunk 1]
```diff
@@ -138,4 +138,40 @@ defmodule KitchenSink.Map do
   def rename_key(_,_,_) do
     %{}
   end
+
+  def new_nested_map(key_list, value) do
```
Comment (kenips) [inline lib/kitchen-sink/map.ex]: how about calling this `make_with_nested_keys` and add @doc.

File: lib/kitchen-sink/map.ex [hunk 2]
```diff
@@ -138,4 +138,40 @@ defmodule KitchenSink.Map do
...
+    key_list
+    |> Enum.reverse
+    |> Enum.reduce(value, fn (key, acc) -> %{key => acc} end)
+    |> Map.new
+  end
+
+  defp renamed_key(map) do
```
Comment (kenips) [inline lib/kitchen-sink/map.ex]: call this `do_rename_key`

File: lib/kitchen-sink/map.ex [hunk 3]
```diff
@@ -138,4 +138,40 @@ defmodule KitchenSink.Map do
...
+        {root_key, new_nested_map(key_list, value)}
+
+      ({old_key, new_key}) ->
+        value = Map.get(map, old_key)
+        {new_key, value}
+    end
+  end
+
+  @doc """
+  renames many keys in a map
```
Comment (kenips) [inline lib/kitchen-sink/map.ex]: Need more detailed doc. Probably good to use doctest here too for transformational functions.

File: lib/kitchen-sink/map.ex [hunk 4]
```diff
@@ -138,4 +138,40 @@ defmodule KitchenSink.Map do
...
+  """
+  def rename_keys(map, key_map, :prune) do
+    key_map
+    |> Enum.map(renamed_key(map))
+    |> Map.new
+  end
+  def rename_keys(map, key_map) do
```
Comment (kenips) [inline lib/kitchen-sink/map.ex]: do not group functions that are not of same arity. You probably meant to pass prune as an option, then make it so as pass it as an option with default = false?

**Review by kenips — CHANGES_REQUESTED** [review]

File: lib/kitchen-sink/map.ex (hunk 1 above)
Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: this created a map with exactly 1 value

Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: new_nested_map is close to the Map.new name.
maybe KMap.new_nested. or KMap.nested?
i don't like the "make" prefix as i don't know of other examples using this prefix

File: lib/kitchen-sink/map.ex (hunk 4 above)
Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: actually, i mean to make it a 3rd optional arg

File: lib/kitchen-sink/map.ex (hunk 1 above)
Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: changed to new_nested

File: lib/kitchen-sink/map.ex (hunk 2 above)
Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: fixed

File: lib/kitchen-sink/map.ex (hunk 3 above)
Comment (boxxxie) [reply lib/kitchen-sink/map.ex]: added

File: lib/kitchen-sink/map.ex [hunk 5]
```diff
@@ -138,4 +138,58 @@ defmodule KitchenSink.Map do
...
+  ```
+
+  If any of the values of the `key_map` are lists, then the output for that key will be a nested map based on the items
+  in the list.
+
+  The third argument is optional, if you use :prune then the Map returned will only contain the keys that are in the
+  ``key_map``
+
+  """
+  def rename_keys(map, key_map, :prune) do
```
Comment (olafura) [inline lib/kitchen-sink/map.ex]: Maybe remap_keys ?

**Review by olafura — CHANGES_REQUESTED** [review]

**Review by olafura — APPROVED** [review]

**Comment by boxxxie** [comment]
@kenips more more

**Review by kenips — APPROVED** [review]

**Review by kenips — APPROVED** [review]

**Pull request merged**